test/discr/discr_ops_test.py

```diff
@@ -0,0 +1,287 @@
+    assert all_equal(df2[1:-1], dfe[1:-1])
+    assert not df1[0] == df2[0]
+    assert not df1[-1] == df2[-1]
+
+    # in-place evaluation
+    out = np.zeros_like(f)
+    assert out is finite_diff(f, out)
+    assert all_equal(out, finite_diff(f))
+    assert out is not finite_diff(f)
+
+    # axis
+    f = np.array([[0., 1., 2., 3., 4.],
+                  [1., 2., 3., 4., 5.]])
+    df0 = finite_diff(f, axis=0)
+    df1 = finite_diff(f, axis=1)
+    assert all_equal(df0, df1)
+
+    # complex arrays
+    f = np.array([0., 1., 2., 3., 4.]) + 1j * np.array([10., 9., 8., 7., 6.])
+    df = finite_diff(f)
+    assert all(df.real == 1)
+    assert all(df.imag == -1)
+
+
+def test_discr_part_deriv():
+    """Discretized partial derivative."""
+
+    discr_space = Rn(1)
+    with pytest.raises(TypeError):
+        DiscretePartDeriv(discr_space)
+
+    # phantom data
+    data = np.array([[1.2, 0, 3, 5, 7],
+                     [4, -1, 3, -2.1, -4]])
+
+    # discretized space
+    space = FunctionSpace(Rectangle([0, 0], [2, 1]))
+    discr_space = uniform_discr_fromspace(space, data.shape)
+
+    # operator
+    par_div = DiscretePartDeriv(discr_space)
+
+    # discretized space vector
+    f = par_div.domain.element(data)
+
+    # partial derivative
+    par_div_f1 = par_div(f)
+
+    # operator
+    par_div = DiscretePartDeriv(discr_space, axis=1, dx=0.2, edge_order=2,
+                                zero_padding=True)
+
+    # partial derivative
+    par_div_f2 = par_div(f)
+
+    assert not par_div_f1 == par_div_f2
+
+
+def ndvolume(lin_size, ndim, dtype=np.float64):
+    """Hypercube phantom.
+
+    Parameters
+    ----------
+    lin_size : `int`
+       Size of array in each dimension
+    ndim : `int`
+        Number of dimensions
+    dtype : dtype
+        The type of the output array
+
+    """
+    s = [1]
+    vol = np.arange(lin_size, dtype=dtype)
+    for _ in range(ndim - 1):
+        s.insert(0, lin_size)
+        vol = vol * vol.reshape(s)
+    return vol
+
+
+def test_discrete_gradient():
+    """Discretized spatial gradient operator."""
+
+    discr_space = Rn(1)
+    with pytest.raises(TypeError):
+        DiscreteGradient(discr_space)
+
+    for ndim in range(1, 6):
+        # DiscreteLp Vector
+        vsize = 3
+        intvl = IntervalProd([0.] * ndim, [vsize] * ndim)
+        space = FunctionSpace(intvl)
+        discr_space = uniform_discr_fromspace(space, [vsize] * ndim)
+        dom_vec = discr_space.element(ndvolume(vsize, ndim))
+
+        # Gradient
+        grad = DiscreteGradient(discr_space)
+        grad(dom_vec)
+
+    # DiscreteLp Vector
+    ndim = 2
+    vsize = 3
+    intvl = IntervalProd([0.] * ndim, [vsize] * ndim)
+    space = FunctionSpace(intvl)
+    discr_space = uniform_discr_fromspace(space, [vsize] * ndim)
+    dom_vec = discr_space.element(ndvolume(vsize, ndim))
+
+    # Gradient
+    grad = DiscreteGradient(discr_space)
+    grad_vec = grad(dom_vec)
+    assert len(grad_vec) is ndim
```

Comment:
Do not compare integers using `is`, see [stockoverflow](http://stackoverflow.com/questions/306313/pythons-is-operator-behaves-unexpectedly-with-integers) for more info.